In [60]:
import random
import numpy as np
import matplotlib.pyplot as plt
import timeit #time to process
import requests
from datetime import datetime

from config import weather_api_key
from citipy import citipy # Use the citipy module to determine city based on latitude and longitude.

In [20]:
requests.__version__
#https://requests.kennethreitz.org/en/master/


'2.24.0'

In [21]:
x = 1
latitudes = []
while x < 11:
    random_lat = random.randint(-90, 89) + random.random()
    latitudes.append(random_lat)
    x += 1
latitudes

[-84.44702953201985,
 7.914911382087724,
 31.355193325787376,
 -89.38178766268466,
 -78.04511902748966,
 27.5465611584234,
 -86.79055346178166,
 -72.52828772051598,
 32.89316011244349,
 12.106682226061421]

In [22]:
#can not generate whole numbers
random.uniform(-90, 90)

-28.99445936449778

In [23]:
# only generate whole numbers
# step is used to seperate the random numbers by a minimum amount
random.randrange(-90, 90, step=3)

-24

In [24]:
#random whole numbers from with in a given range
random.randrange(-90, 90, step=1)

62

In [25]:
#random whole numbers from with in a given range
random.randint(-90,90)

81

In [26]:
# generate and array of x (size) amount of numbers
np.random.uniform(-90.000, 90.000, size=50)

array([  1.09739375,  43.30809485,  88.75797722,  -9.50092803,
        -7.76261415, -38.57737564,  74.17209557,  16.50842538,
       -27.08869747,  89.45852007,   4.08364249,  65.81874268,
       -37.47979464,  82.48684943, -10.47690815, -65.97601785,
        17.90085712,  59.38164088,  43.3381235 ,  40.89478539,
         9.86008449, -44.51550797,  18.06263199,  71.89179015,
       -52.4958868 , -77.56129874,  55.61236614,   9.45574242,
         4.44317369, -15.06969075, -77.17397824,  81.52569539,
         2.41878827,  -2.98067227, -36.59746685,  71.83051681,
        88.06294814,  44.61519231, -73.1754895 , -29.86317298,
       -12.93627968,  33.15045078,  62.87846812, -34.320833  ,
        88.99659652,  17.20543804, -22.5747996 ,  35.819973  ,
       -67.44805196,  -4.17863631])

In [27]:
# time to process line
%timeit np.random.uniform(-90.000, 90.000, size=1500)

17 µs ± 1.24 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [28]:
# For loop take 60 times longer than the np.random.uniform()
def latitudes(size):
    latitudes = []
    x = 0
    while x < (size):
        random_lat = random.randint(-90, 90) + random.random()
        latitudes.append(random_lat)
        x += 1
    return latitudes
# Call the function with 1500.
%timeit latitudes(1500)

1.82 ms ± 152 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


# Generate Random Lat Long

In [29]:
# Create a set of random latitude and longitude combinations.
# zip creates pairs in the order listed
lats = np.random.uniform(low=-90.000, high=90.000, size=1500)
lngs = np.random.uniform(low=-180.000, high=180.000, size=1500)
lat_lngs = zip(lats, lngs)
lat_lngs

#### Example --how to show contents of zip-- DO NOT DO THIS WITH 1500
#### You can only unzip a zipped tuple once before it is removed from the
#### computer's memory. Make sure you unzip the latitudes and longitudes 
#### into the coordinates list before moving on.

In [30]:
# Create a practice set of random latitude and longitude combinations.
x = [25.12903645, 25.92017388, 26.62509167, -59.98969384, 37.30571269]
y = [-67.59741259, 11.09532135, 74.84233102, -76.89176677, -61.13376282]
coordinates = zip(x, y)

# Use the tuple() function to display the latitude and longitude combinations.
for coordinate in coordinates:
    print(coordinate[0], coordinate[1])

25.12903645 -67.59741259
25.92017388 11.09532135
26.62509167 74.84233102
-59.98969384 -76.89176677
37.30571269 -61.13376282


In [31]:
# Add the latitudes and longitudes to a list.
coordinates = list(lat_lngs)

In [32]:
coordinates[25]

(-78.78743823831408, 146.75141908789516)

# Generate Random World Cities
##### Example for iterating through lat long  and finding city

In [33]:
# Use the tuple() function to display the latitude and longitude combinations.
# It works...commenting out to save memory
#for coordinate in coordinates:
#    print(citipy.nearest_city(coordinate[0], coordinate[1]).city_name,
#          citipy.nearest_city(coordinate[0], coordinate[1]).country_code)

In [34]:
# Create a list for holding the cities.
cities = []
# Identify the nearest city for each latitude and longitude combination.
for coordinate in coordinates:
    city = citipy.nearest_city(coordinate[0], coordinate[1]).city_name

    # If the city is unique, then we will add it to the cities list.
    if city not in cities:
        cities.append(city)
# Print the city count to confirm sufficient count.
len(cities)

603

# Using API
### Open Weather map API
### import requests 
##### https://requests.kennethreitz.org/en/master/

### Building API URL

In [35]:
# api.openweathermap.org/data/2.5/weather?q={city name}&appid={API key}
# api.openweathermap.org/data/2.5/weather?q={city name},{state code},{country code}&appid={API key}
# api.openweathermap.org/data/2.5/weather?q={city name},{state code}&appid={API key}


In [36]:
# Starting URL for Weather Map API Call.
url = "http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=" + weather_api_key
print(url)

http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=72c6ac49835b8273bfc5a12e4f65f701


### Add query to url

In [37]:
# Create an endpoint URL for a city.
city_url = url + "&q=" + "Boston"
print(city_url)

http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=72c6ac49835b8273bfc5a12e4f65f701&q=Boston


## Request()
### Request Method	    Action
### get()--------------------->Retrieves data from a web source.
### head()--------------------->Retrieves the headers from a web source.
### post()--------------------->Adds or annotates data on a web source. Used on mailing groups, message boards, or comments.
### put()--------------------->Updates an existing resource. For example, if the date on a Wikipedia page is wrong, you can use the put() method to update that date.
### delete()--------------------->Deletes data from a web source.
### options()--------------------->Discovers what HTTP methods a web source allows.
### patch()--------------------->Partially modifies a web source.


### 
### 
### Build a get request

In [43]:
city_weather = requests.get(city_url)
city_weather

<Response [200]>

In [44]:
#boston spelled wrong so it could not find that name in the query
city_url = url + "&q=" + "Bston"
city_weather = requests.get(city_url)
city_weather

<Response [404]>

In [45]:
city_url

'http://api.openweathermap.org/data/2.5/weather?units=Imperial&APPID=72c6ac49835b8273bfc5a12e4f65f701&q=Bston'

In [46]:
city_url = url + "&q=" + "Boston"
city_weather = requests.get(city_url)
city_weather

<Response [200]>

### Get the data/text from the 'Get' request.
#### Output is a dictionary

In [47]:
# Get the text of the 'Get' request. Output is a dictionary
city_weather.text

'{"coord":{"lon":-71.0598,"lat":42.3584},"weather":[{"id":804,"main":"Clouds","description":"overcast clouds","icon":"04d"}],"base":"stations","main":{"temp":58.95,"feels_like":58.3,"temp_min":56.66,"temp_max":61.18,"pressure":1019,"humidity":80},"visibility":10000,"wind":{"speed":5.01,"deg":329,"gust":14},"clouds":{"all":100},"dt":1622500222,"sys":{"type":2,"id":2013408,"country":"US","sunrise":1622452234,"sunset":1622506423},"timezone":-14400,"id":4930956,"name":"Boston","cod":200}'

### Use json to access the data returned from get request
### Makes it a lot easier to read

In [48]:
# Get the JSON text of the 'Get' request.
city_weather.json()

{'coord': {'lon': -71.0598, 'lat': 42.3584},
 'weather': [{'id': 804,
   'main': 'Clouds',
   'description': 'overcast clouds',
   'icon': '04d'}],
 'base': 'stations',
 'main': {'temp': 58.95,
  'feels_like': 58.3,
  'temp_min': 56.66,
  'temp_max': 61.18,
  'pressure': 1019,
  'humidity': 80},
 'visibility': 10000,
 'wind': {'speed': 5.01, 'deg': 329, 'gust': 14},
 'clouds': {'all': 100},
 'dt': 1622500222,
 'sys': {'type': 2,
  'id': 2013408,
  'country': 'US',
  'sunrise': 1622452234,
  'sunset': 1622506423},
 'timezone': -14400,
 'id': 4930956,
 'name': 'Boston',
 'cod': 200}

### Error handling for request respnse code (200, 404 etc)
### Use .status_code on the variable containing the response

In [51]:
# Create an endpoint URL for a city.
city_url = url + "&q=" + "Bton"
city_weather = requests.get(city_url)
if city_weather.status_code == 200:
    print(f"City Weather found.")
else:
    print(f"City weather not found.")

City weather not found.


In [52]:
# Create an endpoint URL for a city.
city_url = url + "&q=" + "Boston"
city_weather = requests.get(city_url)
if city_weather.status_code == 200:
    print(f"City Weather found.")
else:
    print(f"City weather not found.")

City Weather found.


# PARSE RESPONSE from Get request

### Get the json data

In [ ]:
# Get the JSON data.
boston_data = city_weather.json()

### Review data

In [55]:
# Review data in table
boston_data

{'coord': {'lon': -71.0598, 'lat': 42.3584},
 'weather': [{'id': 804,
   'main': 'Clouds',
   'description': 'overcast clouds',
   'icon': '04d'}],
 'base': 'stations',
 'main': {'temp': 58.98,
  'feels_like': 58.3,
  'temp_min': 56.66,
  'temp_max': 61.18,
  'pressure': 1019,
  'humidity': 79},
 'visibility': 10000,
 'wind': {'speed': 5.01, 'deg': 306, 'gust': 13},
 'clouds': {'all': 100},
 'dt': 1622501306,
 'sys': {'type': 2,
  'id': 2013408,
  'country': 'US',
  'sunrise': 1622452234,
  'sunset': 1622506423},
 'timezone': -14400,
 'id': 4930956,
 'name': 'Boston',
 'cod': 200}

### Pull data from dictionary

In [54]:
boston_data["sys"]

{'type': 2,
 'id': 2013408,
 'country': 'US',
 'sunrise': 1622452234,
 'sunset': 1622506423}

### Pull data from dictionary 2 levels deep

In [56]:
boston_data["sys"]["country"]

'US'

In [57]:
# date time 
boston_data['dt']

1622501306

In [58]:
# latitude
boston_data["coord"]["lat"]

42.3584

### Place all relevent data into variables

In [59]:
lat = boston_data["coord"]["lat"]
lng = boston_data["coord"]["lon"]
max_temp = boston_data["main"]["temp_max"]
humidity = boston_data["main"]["humidity"]
clouds = boston_data["clouds"]["all"]
wind = boston_data["wind"]["speed"]
print(lat, lng, max_temp, humidity, clouds, wind)

42.3584 -71.0598 61.18 79 100 5.01


## Convert Date Time Stamp from unix time
### from datetime import datetime
#### https://docs.python.org/3.6/library/datetime.html#strftime-and-strptime-behavior
#### https://docs.python.org/3.6/library/datetime.html
#### Date time listed in a comma delimited fashion

In [61]:
# Get the date from the JSON file.
date = boston_data["dt"]
# Convert the UTC date to a date format with year, month, day, hours, minutes, and seconds.
datetime.utcfromtimestamp(date)

datetime.datetime(2021, 5, 31, 22, 48, 26)

In [64]:
# Get the date from the JSON file.
date = boston_data["dt"]
# Convert the UTC date to a date string formatted
datetime.utcfromtimestamp(date).strftime('%Y-%m-%d %H:%M:%S')

'2021-05-31 22:48:26'

### Import Dependencies, and Initialize an Empty List and Counters

In [65]:
# Create an empty list to hold the weather data.
city_data = []
# Print the beginning of the logging.
print("Beginning Data Retrieval     ")
print("-----------------------------")

# Create counters.
record_count = 1
set_count = 1

Beginning Data Retrieval     
-----------------------------


### Basic Loop Through the List of Cities and Build the City URL

In [66]:
# Loop through all the cities in our list.
#for i in range(len(cities)):
#
#
#    # Group cities in sets of 50 for logging purposes.
#   if (i % 50 == 0 and i >= 50):
#        set_count += 1
#        record_count = 1
#    # Create endpoint URL with each city.
#    city_url = url + "&q=" + cities[i]

## enumerate() 
### Loop Through the List of Cities and Build the City URL
### it works just not complex enough for what we need


In [71]:
# Loop through all the cities in the list.
for i, city in enumerate(cities):

    # Group cities in sets of 50 for logging purposes.
    if (i % 50 == 0 and i >= 50): #is remainder of 50 >= 0... basically checks if there is a remainder. if so contiue counting until there isn't
        set_count += 1
        record_count = 1
    # Create endpoint URL with each city.
    city_url = url + "&q=" + city.replace(" ","+") # replace all spaces in city with "+"

    # Log the URL, record, and set numbers and the city.
    print(f"Processing Record {record_count} of Set {set_count} | {city}")
    # Add 1 to the record count.
    record_count += 1
    

Processing Record 4 of Set 49 | biscarrosse
Processing Record 5 of Set 49 | jamestown
Processing Record 6 of Set 49 | nikolskoye
Processing Record 7 of Set 49 | kahului
Processing Record 8 of Set 49 | chokurdakh
Processing Record 9 of Set 49 | devgarh
Processing Record 10 of Set 49 | hobyo
Processing Record 11 of Set 49 | hilo
Processing Record 12 of Set 49 | kapaa
Processing Record 13 of Set 49 | dikson
Processing Record 14 of Set 49 | severo-kurilsk
Processing Record 15 of Set 49 | yellowknife
Processing Record 16 of Set 49 | barentsburg
Processing Record 17 of Set 49 | fortuna
Processing Record 18 of Set 49 | lagoa
Processing Record 19 of Set 49 | victoria
Processing Record 20 of Set 49 | kaitangata
Processing Record 21 of Set 49 | taolanaro
Processing Record 22 of Set 49 | abu kamal
Processing Record 23 of Set 49 | tuktoyaktuk
Processing Record 24 of Set 49 | axim
Processing Record 25 of Set 49 | santa rosa
Processing Record 26 of Set 49 | katsiveli
Processing Record 27 of Set 49 |

# Handle API Request Errors with try-except Blocks
### Exception handling

In [72]:
# Loop through all the cities in the list.
for i, city in enumerate(cities):

    # Group cities in sets of 50 for logging purposes.
    if (i % 50 == 0 and i >= 50): #is remainder of 50 >= 0... basically checks if there is a remainder. if so contiue counting until there isn't
        set_count += 1
        record_count = 1
    # Create endpoint URL with each city.
    city_url = url + "&q=" + city.replace(" ","+") # replace all spaces in city with "+"

    # Log the URL, record, and set numbers and the city.
    print(f"Processing Record {record_count} of Set {set_count} | {city}")
    # Add 1 to the record count.
    record_count += 1
    ###################################################################
    #   Start your exception here...when it adds to the iterator      #
    #   to go to the next record                                      #
    ###################################################################
# Run an API request for each of the cities.
    try:
        # Parse the JSON and retrieve data.
        city_weather = requests.get(city_url).json()
        # Parse out the needed data.
        city_lat = city_weather["coord"]["lat"]
        city_lng = city_weather["coord"]["lon"]
        city_max_temp = city_weather["main"]["temp_max"]
        city_humidity = city_weather["main"]["humidity"]
        city_clouds = city_weather["clouds"]["all"]
        city_wind = city_weather["wind"]["speed"]
        city_country = city_weather["sys"]["country"]
        # Convert the date to ISO standard.
        city_date = datetime.utcfromtimestamp(city_weather["dt"]).strftime('%Y-%m-%d %H:%M:%S')
        # Append the city information into city_data list.
        city_data.append({"City": city.title(),
                          "Lat": city_lat,
                          "Lng": city_lng,
                          "Max Temp": city_max_temp,
                          "Humidity": city_humidity,
                          "Cloudiness": city_clouds,
                          "Wind Speed": city_wind,
                          "Country": city_country,
                          "Date": city_date})

# If an error is experienced, skip the city.
    except:
        print("City not found. Skipping...")
        pass

# Indicate that Data Loading is complete.
print("-----------------------------")
print("Data Retrieval Complete      ")
print("-----------------------------")

Processing Record 4 of Set 61 | biscarrosse
Processing Record 5 of Set 61 | jamestown
Processing Record 6 of Set 61 | nikolskoye
Processing Record 7 of Set 61 | kahului
Processing Record 8 of Set 61 | chokurdakh
Processing Record 9 of Set 61 | devgarh
Processing Record 10 of Set 61 | hobyo
Processing Record 11 of Set 61 | hilo
Processing Record 12 of Set 61 | kapaa
Processing Record 13 of Set 61 | dikson
Processing Record 14 of Set 61 | severo-kurilsk
Processing Record 15 of Set 61 | yellowknife
Processing Record 16 of Set 61 | barentsburg
City not found. Skipping...
Processing Record 17 of Set 61 | fortuna
Processing Record 18 of Set 61 | lagoa
Processing Record 19 of Set 61 | victoria
Processing Record 20 of Set 61 | kaitangata
Processing Record 21 of Set 61 | taolanaro
City not found. Skipping...
Processing Record 22 of Set 61 | abu kamal
Processing Record 23 of Set 61 | tuktoyaktuk
Processing Record 24 of Set 61 | axim
Processing Record 25 of Set 61 | santa rosa
Processing Record 2

Processing Record 40 of Set 64 | grand river south east
City not found. Skipping...
Processing Record 41 of Set 64 | banda aceh
Processing Record 42 of Set 64 | kibuye
Processing Record 43 of Set 64 | santa eulalia del rio
City not found. Skipping...
Processing Record 44 of Set 64 | esperance
Processing Record 45 of Set 64 | shimoda
Processing Record 46 of Set 64 | north bend
Processing Record 47 of Set 64 | upernavik
Processing Record 48 of Set 64 | yarmouth
Processing Record 49 of Set 64 | calvia
Processing Record 50 of Set 64 | anderson
Processing Record 1 of Set 65 | avarua
Processing Record 2 of Set 65 | marsa matruh
Processing Record 3 of Set 65 | meulaboh
Processing Record 4 of Set 65 | tomatlan
Processing Record 5 of Set 65 | pangkalanbuun
Processing Record 6 of Set 65 | jaru
Processing Record 7 of Set 65 | puerto escondido
Processing Record 8 of Set 65 | muros
Processing Record 9 of Set 65 | hami
Processing Record 10 of Set 65 | panama city
Processing Record 11 of Set 65 | cor

Processing Record 26 of Set 68 | ilhabela
Processing Record 27 of Set 68 | halifax
Processing Record 28 of Set 68 | port hardy
Processing Record 29 of Set 68 | san cristobal
Processing Record 30 of Set 68 | marienburg
Processing Record 31 of Set 68 | anloga
Processing Record 32 of Set 68 | waingapu
Processing Record 33 of Set 68 | prabumulih
Processing Record 34 of Set 68 | baykit
Processing Record 35 of Set 68 | posse
Processing Record 36 of Set 68 | zdvinsk
Processing Record 37 of Set 68 | kisujszallas
Processing Record 38 of Set 68 | saldanha
Processing Record 39 of Set 68 | el buen pastor
City not found. Skipping...
Processing Record 40 of Set 68 | mackay
Processing Record 41 of Set 68 | bakel
Processing Record 42 of Set 68 | trat
Processing Record 43 of Set 68 | skagen
Processing Record 44 of Set 68 | broome
Processing Record 45 of Set 68 | prieska
Processing Record 46 of Set 68 | bagdarin
Processing Record 47 of Set 68 | iberia
Processing Record 48 of Set 68 | tiarei
Processing R

Processing Record 16 of Set 72 | trofors
Processing Record 17 of Set 72 | florenville
Processing Record 18 of Set 72 | taian
Processing Record 19 of Set 72 | kysyl-syr
Processing Record 20 of Set 72 | mpika
Processing Record 21 of Set 72 | luebo
Processing Record 22 of Set 72 | orlovskiy
Processing Record 23 of Set 72 | clifton
Processing Record 24 of Set 72 | ijaki
City not found. Skipping...
Processing Record 25 of Set 72 | srem
Processing Record 26 of Set 72 | namanga
Processing Record 27 of Set 72 | usinsk
Processing Record 28 of Set 72 | taoudenni
Processing Record 29 of Set 72 | vallenar
Processing Record 30 of Set 72 | dwarka
Processing Record 31 of Set 72 | aktash
Processing Record 32 of Set 72 | roccastrada
Processing Record 33 of Set 72 | warman
Processing Record 34 of Set 72 | guaira
Processing Record 35 of Set 72 | viru
Processing Record 36 of Set 72 | lengshuitan
Processing Record 37 of Set 72 | san quintin
Processing Record 38 of Set 72 | shelburne
Processing Record 39 of